src/data/Rank.ts

```diff
@@ -0,0 +1,168 @@
+}
+
+export function parse_rank(rank_string: string): Rank | void {
+    let re = /\s*([0-9]+)\s*(K(yu)?|D(an)?|P(ro)?)\s*/iy;
+    let result = re.exec(rank_string);
+
+    if (!result) { return; }
+
+    let [_, level_string, type_string] = result;
+    let level = parseInt(level_string, 10);
+    let type = { "k": "Kyu", "d": "Dan", "p": "Pro" }[type_string[0].toLowerCase()];
+
+    if (isNaN(level) || !type) { return; }
+
+    return validate_rank({level: level, type: type});
+}
+
+
+
+// Rank arithmetic: addition, subtraction, equality and constraint.
+export function add_rank(rank: AmateurRank, amount: number): AmateurRank;
+export function add_rank(rank: ProfessionalRank, amount: number): ProfessionalRank;
+export function add_rank(rank: Rank, amount: number): Rank;
+export function add_rank(rank: Rank, amount: number): Rank {
+    rank = validate_rank(rank);
+    amount = Math.floor(amount);
+
+    if (rank.type === "Kyu") {
+        rank.level -= amount;
+    }
+    else {
+        rank.level += amount;
+    }
+
+    return validate_rank(rank);
+}
+
+export function subtract_rank(a: Rank, b: Rank): number {
+    a = validate_rank(a);
+    b = validate_rank(b);
+
+    if (is_professional(a) && is_professional(b)) {
+        return a.level - b.level;
+    }
+    else if (is_professional(a)) {
+        return +Infinity;
+    }
+    else if (is_professional(b)) {
+        return -Infinity;
+    }
+    else {
+        return (a.type === "Dan" ? a.level : 1 - a.level) - (b.type === "Dan" ? b.level : 1 - b.level);
+    }
+}
+
+export function equal_rank(a: Rank, b: Rank): boolean {
+    a = validate_rank(a);
+    b = validate_rank(b);
+
+    return a.type === b.type && a.level === b.level;
+}
+
+export function constrain_rank(lower: Rank, middle: Rank, upper: Rank): Rank {
+    if (subtract_rank(lower, middle) > 0) {
+        return lower;
+    }
+    if (subtract_rank(middle, upper) > 0) {
+        return upper;
+    }
+    return middle;
+}
+
+
+
+// Rank validation. Ranks have integer levels, have level at least 1, and
+// have a set maximum level that depends on the type.
+const maximum_level: {readonly [type: string]: number} = {
+    "Kyu": 30,
+    "Dan": 7,
+    "Pro": 9,
+};
+const adjust_type: {readonly [type: string]: (level: number) => Rank} = {
+    "Kyu": (level) => ({level: 1 - level, type: "Dan"}),
+    "Dan": (level) => ({level: 1 - level, type: "Kyu"}),
+    "Pro": (level) => ({level: 1,         type: "Pro"})
+};
+function validate_rank(rank: AmateurRank): AmateurRank;
+function validate_rank(rank: ProfessionalRank): ProfessionalRank;
+function validate_rank(rank: Rank): Rank;
+function validate_rank(rank: Rank): Rank {
+    let level: number = rank.level;
+    level = Math.floor(level);
+    level = Math.min(level, maximum_level[rank.type]);
+    if (level < 1) {
+        return adjust_type[rank.type](level);
+    }
+    else {
+        return {level: level, type: rank.type};
+    }
+}
```

Comment:
I gotta be honest, the rank updates you have here seem like they've complicated things a lot.. it might just take some getting used to, so I'll try, but so far it seems like it's introduced a lot of code and complexity and I'm not sure I feel the gain yet.